src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,186 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        final Node<T> newTail = tail;
+        Node<T> newNode = new Node<>(tail, value, null);
+        tail = newNode;
+        if (newTail == null) {
+            head = newNode;
+        } else {
+            newTail.next = tail;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkPositionIndex(index);
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> oldAtIndex = head;
+            for (int i = 0; i < index; i++) {
+                oldAtIndex = oldAtIndex.next;
+            }
+            final Node<T> newNode = new Node<>(oldAtIndex.prev, value, oldAtIndex);
+            if (oldAtIndex.prev == null) {
+                head = newNode;
+            } else {
+                oldAtIndex.prev.next = newNode;
+            }
+            oldAtIndex.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        Object[] listArray = list.toArray();
+        final int numNew = listArray.length;
+        Node<T> pred;
+        Node<T> succ;
+        succ = null;
+        pred = tail;
+        for (Object object : listArray) {
+            T element = (T) object;
+            add(element);
+        }
```

Comment:
```suggestion
        for (T element : list) {
            add(element);
        }
```
You can simplify this method